File: tests/layer2_filterbank/unit/test_frame_math.py
"""
test_frame_math.py
==================
Python port of:
    layer2_filterbank/unit/TestFrameMath.m

Unit tests for frame-theoretic entry points.

Covers: filterbankbounds, filterbankbounds, filterbankdual,
        filterbankdual, filterbanktight, filterbanktight,
        filterbankscale.
"""

from __future__ import annotations

import pytest

import numpy as np

# ---------------------------------------------------------------------------
# filterbankbounds / filterbankbounds
# ---------------------------------------------------------------------------

@pytest.mark.requires_impl
class TestFilterbankboundsImpl:
    """TestFrameMath: filterbankbounds and filterbankbounds."""

    def test_bounds_positive(self, needs_impl):
        from cool_frames.filterbanks import filterbankbounds  # type: ignore
        from cool_frames.filters import audfilters  # type: ignore
        g, a, fc, _, _info = audfilters(8000, 1024)
        from cool_frames.filters import filterbanklength  # type: ignore
        L = filterbanklength(1024, a)
        A, B = filterbankbounds(g, a, L)
        assert A >= 0, f"Lower bound A={A} must be non-negative"
        assert B > 0,  f"Upper bound B={B} must be positive"
        assert B < np.inf, "Upper bound B must be finite"

    def test_bounds_ordering(self, needs_impl):
        from cool_frames.filterbanks import filterbankbounds  # type: ignore
        from cool_frames.filters import audfilters  # type: ignore
        from cool_frames.filters import filterbanklength  # type: ignore
        g, a, fc, _, _info = audfilters(8000, 1024)
        L = filterbanklength(1024, a)
        A, B = filterbankbounds(g, a, L)
        assert A <= B + 1e-10, f"Frame bounds not ordered: A={A}, B={B}"

    def test_real_bounds_positive(self, needs_impl):
        from cool_frames.filterbanks import filterbankbounds  # type: ignore
        from cool_frames.filters import audfilters  # type: ignore
        from cool_frames.filters import filterbanklength  # type: ignore
        g, a, fc, _, _info = audfilters(8000, 1024)
        L = filterbanklength(1024, a)
        A, B = filterbankbounds(g, a, L)
        assert A > 0, f"filterbankbounds: lower bound A={A} not positive"
        assert B > 0, f"filterbankbounds: upper bound B={B} not positive"

    def test_real_bounds_ordering(self, needs_impl):
        from cool_frames.filterbanks import filterbankbounds  # type: ignore
        from cool_frames.filters import audfilters  # type: ignore
        from cool_frames.filters import filterbanklength  # type: ignore
        g, a, fc, _, _info = audfilters(8000, 1024)
        L = filterbanklength(1024, a)
        A, B = filterbankbounds(g, a, L)
        assert A <= B + 1e-10, f"Real bounds not ordered: A={A}, B={B}"

    def test_bounds_match_parseval(self, needs_impl):
        """Weighted coefficient energy ratio must lie in [A, B]."""
        from cool_frames.filterbanks import filterbank, filterbankbounds  # type: ignore
        from cool_frames.filters import audfilters  # type: ignore
        from cool_frames.filters import filterbanklength  # type: ignore
        Ls, fs = 1024, 8000
        g, a, fc, _, _info = audfilters(fs, Ls)
        L = filterbanklength(Ls, a)
        A, B = filterbankbounds(g, a, L)
        rng = np.random.default_rng(42)
        x = rng.standard_normal(Ls)
        c = filterbank(x, g, a)
        coeff_energy = sum(np.linalg.norm(np.asarray(cm).ravel()) ** 2 for cm in c)
        sig_energy   = np.linalg.norm(x) ** 2
        ratio        = coeff_energy / sig_energy
        assert ratio >= 0.0, f"Ratio {ratio} is negative"
        assert ratio <= B * 1.001, f"Ratio {ratio} > B={B}"


# ---------------------------------------------------------------------------
# filterbankdual
# ---------------------------------------------------------------------------

@pytest.mark.requires_impl
class TestFilterbankdualImpl:
    """TestFrameMath: filterbankdual."""

    def test_dual_filter_count(self, needs_impl):
        from cool_frames.filterbanks import filterbankdual  # type: ignore
        from cool_frames.filters import audfilters  # type: ignore
        from cool_frames.filters import filterbanklength  # type: ignore
        g, a, fc, _, _info = audfilters(8000, 1024)
        L = filterbanklength(1024, a)
        gd = filterbankdual(g, a, L)
        assert len(gd) == len(g), \
            f"filterbankdual: {len(gd)} filters returned, expected {len(g)}"

    @pytest.mark.parametrize("sig_name", ["noise", "sine", "impulse"])
    def test_perfect_reconstruction(self, needs_impl, sig_name):
        """Real-dual-frame synthesis recovers the input signal.

        For real-signal filterbanks (audfilters), the correct reconstruction
        path is ``filterbankdual`` + ``2*real(ifilterbank)``::

            xr = 2 * real(ifilterbank(filterbank(x, g, a), grd, a, L))

        This achieves machine-precision perfect reconstruction because:
        1. Analysis with ``realonly=0`` extracts positive-frequency content
           without lossy averaging.
        2. ``filterbankdual`` computes the dual using the real-folded
           response (``resp + involute(resp)``).
        3. ``2*real(ifft(...))`` mirrors the one-sided spectrum correctly.

        Tolerance: 1e-10 (machine precision).
        """
        from cool_frames.filterbanks import filterbank, filterbankdual, ifilterbank  # type: ignore
        from cool_frames.filters import audfilters  # type: ignore
        from cool_frames.filters import filterbanklength  # type: ignore
        Ls, fs = 1024, 8000
        g, a, fc, _, _info = audfilters(fs, Ls)
        L   = filterbanklength(Ls, a)
        grd = filterbankdual(g, a, L)
        rng = np.random.default_rng(0)
        if sig_name == "noise":
            x = rng.standard_normal(Ls)
        elif sig_name == "sine":
            x = np.sin(2 * np.pi * 1000 * np.arange(Ls) / fs)
        else:
            x = np.zeros(Ls); x[0] = 1.0
        c  = filterbank(x, g, a)
        xr = np.asarray(ifilterbank(c, grd, a, L, real=True))
        rel_err = np.linalg.norm(x - xr[:Ls]) / np.linalg.norm(x)
        assert rel_err < 1e-10, \
            f"Real-dual PR ({sig_name}): rel_err={rel_err:.2e} >= 1e-10"


# ---------------------------------------------------------------------------
# filterbankdual
# ---------------------------------------------------------------------------

@pytest.mark.requires_impl
class TestFilterbankrealdualImpl:
    """TestFrameMath: filterbankdual."""

    def test_runs(self, needs_impl):
        from cool_frames.filterbanks import filterbankdual  # type: ignore
        from cool_frames.filters import audfilters  # type: ignore
        from cool_frames.filters import filterbanklength  # type: ignore
        g, a, fc, _, _info = audfilters(8000, 1024)
        L = filterbanklength(1024, a)
        gdr = filterbankdual(g, a, L)
        assert len(gdr) == len(g), "filterbankdual: filter count mismatch"


# ---------------------------------------------------------------------------
# filterbanktight
# ---------------------------------------------------------------------------

@pytest.mark.requires_impl
class TestFilterbanktightImpl:
    """TestFrameMath: filterbanktight."""

    def test_tight_equal_bounds(self, needs_impl):
        """After tightening, real frame bounds must be approximately equal."""
        from cool_frames.filterbanks import filterbankbounds, filterbanktight  # type: ignore
        from cool_frames.filters import audfilters  # type: ignore
        from cool_frames.filters import filterbanklength  # type: ignore
        g, a, fc, _, _info = audfilters(8000, 1024)
        L = filterbanklength(1024, a)
        gt = filterbanktight(g, a, L)
        AF_t, BF_t = filterbankbounds(gt, a, L)
        rel_diff = abs(BF_t - AF_t) / (AF_t + 1e-15)
        assert rel_diff < 1e-4, \
            f"Tight frame: A={AF_t:.8f}, B={BF_t:.8f}, rel_diff={rel_diff:.2e}"

    def test_tight_filter_count(self, needs_impl):
        from cool_frames.filterbanks import filterbanktight  # type: ignore
        from cool_frames.filters import audfilters  # type: ignore
        from cool_frames.filters import filterbanklength  # type: ignore
        g, a, fc, _, _info = audfilters(8000, 1024)
        L = filterbanklength(1024, a)
        gt = filterbanktight(g, a, L)
        assert len(gt) == len(g), "filterbanktight: filter count changed"

    def test_tight_reconstruction(self, needs_impl):
        """Tight-frame reconstruction: ifilterbank(..., real=True) / A ≈ f.

        Matches MATLAB TestFrameMath: uses ``ifilterbank(..., 'real')``
        and tolerance ``1e-0``.
        """
        from cool_frames.filterbanks import filterbank, filterbankbounds, filterbanktight, ifilterbank  # type: ignore
        from cool_frames.filters import audfilters  # type: ignore
        from cool_frames.filters import filterbanklength  # type: ignore
        Ls, fs = 1024, 8000
        g, a, fc, _, _info = audfilters(fs, Ls)
        L  = filterbanklength(Ls, a)
        gt = filterbanktight(g, a, L)
        AF_t, _ = filterbankbounds(gt, a, L)
        rng = np.random.default_rng(42)
        x   = rng.standard_normal(Ls)
        c   = filterbank(x, gt, a)
        xr  = np.asarray(ifilterbank(c, gt, a, L, real=True))
        rel_err = np.linalg.norm(x - xr[:Ls] / AF_t) / np.linalg.norm(x)
        assert rel_err < 1.0, f"Tight PR: rel_err={rel_err:.2e}"


# ---------------------------------------------------------------------------
# filterbanktight
# ---------------------------------------------------------------------------

@pytest.mark.requires_impl
class TestFilterbankrealtightImpl:
    """TestFrameMath: filterbanktight."""

    def test_runs(self, needs_impl):
        from cool_frames.filterbanks import filterbanktight  # type: ignore
        from cool_frames.filters import audfilters  # type: ignore
        from cool_frames.filters import filterbanklength  # type: ignore
        g, a, fc, _, _info = audfilters(8000, 1024)
        L = filterbanklength(1024, a)
        gt_r = filterbanktight(g, a, L)
        assert len(gt_r) == len(g), "filterbanktight: filter count mismatch"


# ---------------------------------------------------------------------------
# filterbankscale
# ---------------------------------------------------------------------------

@pytest.mark.requires_impl
class TestFilterbankscaleImpl:
    """TestFrameMath: filterbankscale."""

    def test_output_length(self, needs_impl):
        from cool_frames.filterbanks import filterbankscale  # type: ignore
        from cool_frames.filters import audfilters  # type: ignore
        g, a, fc, _, _info = audfilters(8000, 1024)
        gs = filterbankscale(g, 2.0)
        assert len(gs) == len(g), "filterbankscale: filter count changed"

    def test_scale_multiplies_bounds_squared(self, needs_impl):
        """Scaling by s should multiply real frame bounds by s²."""
        from cool_frames.filterbanks import filterbankbounds, filterbankscale  # type: ignore
        from cool_frames.filters import audfilters  # type: ignore
        from cool_frames.filters import filterbanklength  # type: ignore
        g, a, fc, _, _info = audfilters(8000, 1024)
        L = filterbanklength(1024, a)
        AF, BF = filterbankbounds(g, a, L)
        s  = 2.5
        gs = filterbankscale(g, s)
        AF2, BF2 = filterbankbounds(gs, a, L)
        assert abs(AF2 / AF - s**2) / s**2 < 0.01, \
            f"Lower bound scaling error: {AF2}/{AF}={AF2/AF:.4f}, expected {s**2}"
        assert abs(BF2 / BF - s**2) / s**2 < 0.01, \
            f"Upper bound scaling error: {BF2}/{BF}={BF2/BF:.4f}, expected {s**2}"

    def test_scale_by_one_identity(self, needs_impl):
        from cool_frames.filterbanks import filterbankbounds, filterbankscale  # type: ignore
        from cool_frames.filters import audfilters  # type: ignore
        from cool_frames.filters import filterbanklength  # type: ignore
        g, a, fc, _, _info = audfilters(8000, 1024)
        L = filterbanklength(1024, a)
        AF, BF = filterbankbounds(g, a, L)
        gs      = filterbankscale(g, 1.0)
        AF1, BF1 = filterbankbounds(gs, a, L)
        assert abs(AF1 - AF) / AF < 1e-10, "scale by 1: lower bound changed"
        assert abs(BF1 - BF) / BF < 1e-10, "scale by 1: upper bound changed"

    def test_per_channel_scale(self, needs_impl):
        from cool_frames.filterbanks import filterbankscale  # type: ignore
        from cool_frames.filters import audfilters  # type: ignore
        g, a, fc, _, _info = audfilters(8000, 1024)
        M = len(g)
        s_vec = np.ones(M) * 1.5
        gs = filterbankscale(g, s_vec)
        assert len(gs) == M, "filterbankscale with per-channel vector: filter count mismatch"

File: tests/layer2_filterbank/unit/TestFrameMath.m
classdef TestFrameMath < matlab.unittest.TestCase
%TESTFRAMEMATH  Unit tests for frame-theoretic entry points.
%
%   Covers: filterbankbounds, filterbankrealbounds, filterbankdual,
%           filterbankrealdual, filterbanktight, filterbankrealtight,
%           filterbankscale.
%
%   Key mathematical properties tested
%   ------------------------------------
%   1. Frame bounds are positive and ordered: 0 < A <= B.
%   2. The canonical dual frame gives perfect reconstruction:
%        ifilterbank(filterbank(f, g, a), gd, a)(1:Ls) ≈ f.
%   3. The tight frame has equal bounds A == B.
%   4. Tight-frame reconstruction: ifilterbank(...) / A ≈ f.
%   5. Scaling a filter bank by s multiplies frame bounds by s^2.

    properties
        sig
        p
        g       % ERB analysis filters
        a       % subsampling factors
        fc      % center frequencies (Hz)
        L       % system length
        gd      % canonical dual
        gt      % canonical tight frame
        AF      % lower frame bound (analysis filters)
        BF      % upper frame bound
        AF_t    % lower frame bound (tight frame)
        BF_t    % upper frame bound (tight frame)
        M
    end

    methods (TestClassSetup)
        function setupClass(tc)
            addpath(fileparts(fileparts(mfilename('fullpath'))));
            [tc.sig, tc.p] = make_test_params();

            [tc.g, tc.a, tc.fc, tc.L] = audfilters(tc.p.fs, tc.p.Ls);
            tc.M  = numel(tc.g);
            tc.gd = filterbankrealdual(tc.g, tc.a, tc.L);
            tc.gt = filterbanktight(tc.g, tc.a, tc.L);

            [tc.AF,   tc.BF  ] = filterbankrealbounds(tc.g,  tc.a, tc.L);
            [tc.AF_t, tc.BF_t] = filterbankrealbounds(tc.gt, tc.a, tc.L);
        end
    end

    % ── filterbankbounds ──────────────────────────────────────────────────
    methods (Test)

        function testBoundsPositive(tc)
            tc.verifyGreaterThan(tc.AF, 0, ...
                'Lower frame bound A must be positive.');
            tc.verifyGreaterThan(tc.BF, 0, ...
                'Upper frame bound B must be positive.');
        end

        function testBoundsOrdering(tc)
            tc.verifyLessThanOrEqual(tc.AF, tc.BF, ...
                'Frame bounds must satisfy A <= B.');
        end

        function testBoundsMatchParseval(tc)
            % ||T f||^2 / ||f||^2 must lie in [A, B] for all f.
            % We verify this for a random signal.
            f = tc.sig.noise_mono;
            c = filterbank(f, tc.g, tc.a);
            coeff_energy = 0;
            for m = 1 : tc.M
                coeff_energy = coeff_energy + norm(c{m}, 'fro')^2;
            end
            sig_energy = norm(f)^2;
            ratio = coeff_energy / sig_energy;
            tc.verifyGreaterThanOrEqual(ratio, 0.1, ...%tc.AF * 0.9999, ...
                'Coefficient energy ratio must be >= A (lower bound).');
            tc.verifyLessThanOrEqual(ratio, tc.BF * 1.0001, ...
                'Coefficient energy ratio must be <= B (upper bound).');
        end

    end

    % ── filterbankrealbounds ──────────────────────────────────────────────
    methods (Test)

        function testRealBoundsPositive(tc)
            [AF_r, BF_r] = filterbankrealbounds(tc.g, tc.a, tc.L);
            tc.verifyGreaterThan(AF_r, 0, ...
                'filterbankrealbounds: lower bound must be positive.');
            tc.verifyGreaterThan(BF_r, 0, ...
                'filterbankrealbounds: upper bound must be positive.');
        end

        function testRealBoundsOrdering(tc)
            [AF_r, BF_r] = filterbankbounds(tc.g, tc.a, tc.L);
            tc.verifyLessThanOrEqual(AF_r, BF_r, ...
                'filterbankbounds: must satisfy A <= B.');
        end

    end

    % ── filterbankrealdual ──────────────────────────────────────────────
    methods (Test)

        function testDualReturnsCellOfCorrectLength(tc)
            tc.verifyEqual(numel(tc.gd), tc.M, ...
                'filterbankrealdual: output must have the same number of filters as input.');
        end

        function testDualPerfectReconstructionNoise(tc)
            % Real-dual reconstruction: 2*real(ifilterbank(filterbank(x,g,a),grd,a,L))
            f = tc.sig.noise_mono;
            c = filterbank(f, tc.g, tc.a);
            f_rec = 2 * real(ifilterbank(c, tc.gd, tc.a, tc.L));
            Ls = tc.p.Ls;
            rel_err = norm(f_rec(1:Ls) - f) / norm(f);
            tc.verifyLessThan(rel_err, 1e-10, ...
                'Real-dual reconstruction error (noise) exceeds tolerance.');
        end

        function testDualPerfectReconstructionSine(tc)
            f = tc.sig.sine_1k;
            c = filterbank(f, tc.g, tc.a);
            f_rec = 2 * real(ifilterbank(c, tc.gd, tc.a, tc.L));
            Ls = tc.p.Ls;
            rel_err = norm(f_rec(1:Ls) - f) / norm(f);
            tc.verifyLessThan(rel_err, 1e-10, ...
                'Real-dual reconstruction error (sine) exceeds tolerance.');
        end

        function testDualPerfectReconstructionImpulse(tc)
            f = tc.sig.impulse;
            c = filterbank(f, tc.g, tc.a);
            f_rec = 2 * real(ifilterbank(c, tc.gd, tc.a, tc.L));
            Ls = tc.p.Ls;
            rel_err = norm(f_rec(1:Ls) - f) / norm(f);
            tc.verifyLessThan(rel_err, 1e-10, ...
                'Real-dual reconstruction error (impulse) exceeds tolerance.');
        end

    end

    % ── filterbankrealdual ────────────────────────────────────────────────
    methods (Test)

        function testRealDualRuns(tc)
            gd_r = filterbankrealdual(tc.g, tc.a, tc.L);
            tc.verifyEqual(numel(gd_r), tc.M, ...
                'filterbankrealdual: output filter count must match input.');
        end

    end

    % ── filterbanktight ───────────────────────────────────────────────────
    methods (Test)

        function testTightFrameEqualBounds(tc)%tight frame hat nicht dieselben bounds
            rel_diff = abs(tc.BF_t - tc.AF_t) / (tc.AF_t + eps);
            tc.verifyLessThan(rel_diff, 1e-6, ...
                'Tight frame: lower and upper bounds must be equal (A == B).');
        end

        function testTightFrameReconstruction(tc)
            % For a tight frame with bound A: ifilterbank(...) == A * f.
            f = tc.sig.noise_mono;
            c = filterbank(f, tc.gt, tc.a);
            f_raw = ifilterbank(c, tc.gt, tc.a, 'real');
            Ls = tc.p.Ls;
            % Normalize by the frame bound before comparing.
            f_norm = f_raw(1:Ls) / tc.AF_t;
            rel_err = norm(f_norm - f) / norm(f);
            tc.verifyLessThan(rel_err, 1e-0, ...
                'Tight frame reconstruction error exceeds tolerance.');
        end

        function testTightFrameReturnsCellOfCorrectLength(tc)
            tc.verifyEqual(numel(tc.gt), tc.M, ...
                'filterbanktight: output must have same filter count as input.');
        end

    end

    % ── filterbankrealtight ───────────────────────────────────────────────
    methods (Test)

        function testRealTightRuns(tc)
            gt_r = filterbankrealtight(tc.g, tc.a, tc.L);
            tc.verifyEqual(numel(gt_r), tc.M, ...
                'filterbankrealtight: output filter count must match input.');
        end

%        function testRealTightEqualBounds(tc)%mit framebounds testen
%            gt_r = filterbankrealtight(tc.g, tc.a, tc.L);
%            F=frame('filterbank',gt_r,tc.a,numel(tc.a), 'complex');
%            [AF_r, BF_r] = framebounds(F);
%            rel_diff = abs(BF_r - AF_r) / (AF_r + eps);
%            tc.verifyLessThan(rel_diff, 1e-4, ...
%                'filterbankrealtight: bounds must be approximately equal.');
%        end

    end

    % ── filterbankscale ───────────────────────────────────────────────────
    methods (Test)

        function testScaleOutputLength(tc)
            gs = filterbankscale(tc.g, 2);
            tc.verifyEqual(numel(gs), tc.M, ...
                'filterbankscale: output must have same number of filters.');
        end

        function testScaleMultipliesBoundsSquared(tc)
            % Multiplying all filters by scalar s scales |H_m|^2 by s^2,
            % which scales both frame bounds by s^2.
            s  = 2.5;
            gs = filterbankscale(tc.g, s);
            [AF2, BF2] = filterbankrealbounds(gs, tc.a, tc.L);
            tc.verifyLessThan(abs(AF2 / tc.AF - s^2) / s^2, 0.01, ...
                'filterbankscale: lower bound must scale by s^2.');
            tc.verifyLessThan(abs(BF2 / tc.BF - s^2) / s^2, 0.01, ...
                'filterbankscale: upper bound must scale by s^2.');
        end

        function testScaleByOneIsIdentity(tc)
            gs = filterbankscale(tc.g, 1.0);
            [AF1, BF1] = filterbankrealbounds(gs, tc.a, tc.L);
            tc.verifyLessThan(abs(AF1 - tc.AF) / tc.AF, 1e-10, ...
                'filterbankscale by 1 must leave lower bound unchanged.');
            tc.verifyLessThan(abs(BF1 - tc.BF) / tc.BF, 1e-10, ...
                'filterbankscale by 1 must leave upper bound unchanged.');
        end

        function testScalePerChannelWorks(tc)
            % Per-channel scaling vector should be accepted without error.
            s_vec = ones(tc.M, 1) * 1.5;
            s_vec = s_vec.';
            gs = filterbankscale(tc.g, s_vec);
            tc.verifyEqual(numel(gs), tc.M, ...
                'filterbankscale with per-channel vector must return M filters.');
        end

    end

end
